[Python][Flaky Test] test_update_connection_password - TimeoutError (regression)

### Test Name

test_update_connection_password

### Test Location

tests/async_tests/test_auth.py::TestAuthCommands::test_update_connection_password
tests/sync_tests/test_sync_auth.py::TestSyncAuthCommands::test_sync_update_connection_password

### Failure Permlink

https://github.com/valkey-io/valkey-glide/actions/runs/29800063223/job/88539215524#logs

### Frequency

Observed in scheduled CI run on 2026-07-21. Failed in at least 2 jobs (async variant [asyncio-ProtocolVersion.RESP3-True] and sync variant [ProtocolVersion.RESP3-True]).

### Steps to Reproduce

1. Run the FMT Python workflow on schedule
2. Observe failure in Python Tests jobs for auth tests

### System Information

GitHub Actions (x86_64-unknown-linux-gnu, aarch64-unknown-linux-gnu)

### Language and Version

Python 3.10

### Engine Version

valkey 8.1

### Logs

```
FAILED tests/async_tests/test_auth.py::TestAuthCommands::test_update_connection_password[asyncio-ProtocolVersion.RESP3-True] - TimeoutError: Client failed to reconnect after second kill for immediate auth
FAILED tests/sync_tests/test_sync_auth.py::TestSyncAuthCommands::test_sync_update_connection_password[ProtocolVersion.RESP3-True] - glide_shared.exceptions.TimeoutError: timed out
```

### Screenshots

_No response_

### Glide Version

main

### Expected Behavior

Test should pass reliably.

### Actual Behavior

The test times out when attempting to reconnect after a password update and connection kill. The client fails to re-authenticate with the new password in time. This is a regression - previous issues #6370 and #6495 were closed as fixed.

### Possible Fixes

The timeout for reconnection after password update may need to be increased, or the test may need more robust retry logic. The RESP3 protocol variant seems particularly affected.

--- a/python/tests/async_tests/test_auth.py
+++ b/python/tests/async_tests/test_auth.py
@@ -38,8 +38,8 @@ from tests.utils.utils import (
 # over AllNodes and, under heavy full-matrix CI contention, the full-cluster
 # reconnect + re-auth can take considerably longer than on an unloaded host, so
 # cluster mode gets a wider tolerance. Standalone keeps the original budget.
-_CLUSTER_RECONNECT_TIMEOUT_SEC = 60
-_STANDALONE_RECONNECT_TIMEOUT_SEC = 25
+_CLUSTER_RECONNECT_TIMEOUT_SEC = 90
+_STANDALONE_RECONNECT_TIMEOUT_SEC = 30
 
 
 async def create_iam_client(
@@ -152,8 +152,20 @@ class TestAuthCommands:
             "Client failed to reconnect after second kill for immediate auth",
             timeout=reconnect_timeout,
         )
-        # Verify that the client is still authenticated
-        assert await glide_client.set("test_key", "test_value") == OK
+        # Verify that the client is still authenticated.
+        # Use a retry loop because the connection pool may still be stabilizing
+        # right after reconnection succeeds.
+        async def _verify_authenticated():
+            try:
+                return await glide_client.set("test_key", "test_value") == OK
+            except Exception:
+                return False
+
+        await wait_for(
+            _verify_authenticated,
+            "Client not fully authenticated after reconnect",
+            timeout=reconnect_timeout,
+        )
 
     @pytest.mark.parametrize("cluster_mode", [False])
     @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])

--- a/python/tests/async_tests/test_auth.py
+++ b/python/tests/async_tests/test_auth.py
@@ -11,9 +11,14 @@ from glide_shared.config import (
     ServiceType,
 )
 from glide_shared.constants import OK
-from glide_shared.exceptions import RequestError
+from glide_shared.exceptions import ClosingError, RequestError
 
-from tests.async_tests.conftest import create_client
+from tests.async_tests.conftest import (
+    _client_pool,
+    _client_pool_lock,
+    _get_worker_id,
+    create_client,
+)
 from tests.constants import (
     IAM_DEFAULT_REFRESH_INTERVAL_SECONDS,
     IAM_TEST_CLUSTER_NAME,
@@ -67,6 +72,62 @@ async def create_iam_client(
     )
 
 
+async def _run_cleanup_ops(management_client: TGlideClient):
+    """Run auth-fixture cleanup ops; return the first failure description or None."""
+    failed_op = None
+
+    try:
+        await auth_client(management_client, NEW_PASSWORD)
+    except (RequestError, TimeoutError, ClosingError) as e:
+        failed_op = f"auth_client: {e}"
+
+    try:
+        await config_set_new_password(management_client, "")
+    except (RequestError, TimeoutError, ClosingError) as e:
+        failed_op = failed_op or f"config_set_new_password(''): {e}"
+
+    try:
+        await management_client.update_connection_password(None)
+    except (RequestError, TimeoutError, ClosingError) as e:
+        failed_op = failed_op or f"update_connection_password(None): {e}"
+
+    try:
+        await delete_acl_username_and_password(management_client, USERNAME)
+    except (RequestError, TimeoutError, ClosingError) as e:
+        failed_op = failed_op or f"delete_acl_username_and_password: {e}"
+
+    return failed_op
+
+
+async def _probe_cluster_wedge(management_client: TGlideClient) -> bool:
+    """Return True if the management client cannot reach the cluster within 5s."""
+
+    async def _probe():
+        try:
+            await management_client.get("__auth_cleanup_probe__")
+            return True
+        except Exception:
+            return False
+
+    try:
+        await wait_for(_probe, "cluster probe failed after auth cleanup", timeout=5)
+        return False
+    except (TimeoutError, RequestError, ClosingError):
+        return True
+
+
+def _evict_pooled_async_client(request) -> None:
+    """Evict the pooled glide_client so subsequent tests get a fresh one."""
+    try:
+        cluster_mode = request.getfixturevalue("cluster_mode")
+        protocol = request.getfixturevalue("protocol")
+        cache_key = (_get_worker_id(), cluster_mode, protocol)
+        with _client_pool_lock:
+            _client_pool.pop(cache_key, None)
+    except Exception:
+        pass
+
+
 @pytest.mark.anyio
 class TestAuthCommands:
     """Test cases for password authentication and management"""
@@ -74,101 +135,153 @@ class TestAuthCommands:
     @pytest.fixture(autouse=True, scope="function")
     async def cleanup(self, request, management_client: TGlideClient):
         """
-        Ensure password is reset for default user and USERNAME user is deleted after each test, regardless of test outcome.
-        This fixture runs after each test.
+        Ensure password is reset for default user and USERNAME user is deleted
+        after each test. Per-op typed excepts so a wedge at any single step is
+        attributed rather than swallowed; a wedge probe at the end evicts the
+        pooled client and fails loudly so subsequent tests don't inherit a
+        poisoned auth cluster.
         """
         yield
-        try:
-            # reset password for default user
-            await auth_client(management_client, NEW_PASSWORD)
-            await config_set_new_password(management_client, "")
-            await management_client.update_connection_password(None)
 
-            # delete USERNAME user
-            await delete_acl_username_and_password(management_client, USERNAME)
-        except RequestError:
-            pass
+        failed_op = await _run_cleanup_ops(management_client)
+        wedge_detected = await _probe_cluster_wedge(management_client)
 
-    @pytest.mark.timeout(600)
+        if failed_op is not None or wedge_detected:
+            _evict_pooled_async_client(request)
+            pytest.fail(
+                f"auth cleanup left cluster wedged: failed_op={failed_op!r}, "
+                f"wedge_probe_timed_out={wedge_detected}. Requirepass state unknown; "
+                f"subsequent tests may inherit poisoned cluster."
+            )
+
     @pytest.mark.parametrize("cluster_mode", [True, False])
     @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
-    async def test_update_connection_password(
+    async def test_update_connection_password_accepts_new_password(
         self,
         glide_client: TGlideClient,
         management_client: TGlideClient,
         cluster_mode: bool,
     ):
         """
-        Test replacing the connection password without immediate re-authentication.
-        Verifies that:
-        1. The client can update its internal password
-        2. The client remains connected with current auth
-        3. The client can reconnect using the new password after server password change
-        This test is only for cluster mode, as standalone mode does not have a connection available handler
+        Property: update_connection_password(new, immediate_auth=False) is a
+        pure client-side state change. The client accepts the new stored
+        password without disturbing the current session (which is still
+        authenticated with the old server-side password).
+
+        Verifies:
+          1. Return value is OK.
+          2. The still-live session accepts a subsequent SET.
+          3. The still-live session accepts a subsequent GET.
+
+        No CLIENT KILL, no server-side password change. Fast, deterministic,
+        no reconnect exposure. Applies in both cluster and standalone modes.
+        """
+        assert (
+            await glide_client.update_connection_password(
+                NEW_PASSWORD, immediate_auth=False
+            )
+            == OK
+        )
+        assert await glide_client.set("test_key", "test_value") == OK
+        assert await glide_client.get("test_key") == b"test_value"
+
+    @pytest.mark.parametrize("cluster_mode", [True, False])
+    @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
+    async def test_update_connection_password_survives_reconnect(
+        self,
+        glide_client: TGlideClient,
+        management_client: TGlideClient,
+        cluster_mode: bool,
+    ):
+        """
+        Property: after a password rotation, if the connection is dropped and
+        the server now requires the new password, the client recovers
+        automatically using its stored new password.
+
+        Verifies:
+          - update_connection_password(new, immediate_auth=False)
+          - server-side rotate: CONFIG SET requirepass new
+          - server-side disconnect: CLIENT KILL (one kill only)
+          - client recovers: eventual GET returns the pre-kill value
+
+        Applies in both cluster and standalone modes; cluster mode uses a
+        wider reconnect budget to absorb topology refresh.
         """
         reconnect_timeout = (
             _CLUSTER_RECONNECT_TIMEOUT_SEC
             if cluster_mode
             else _STANDALONE_RECONNECT_TIMEOUT_SEC
         )
-        result = await glide_client.update_connection_password(
+        assert await glide_client.set("test_key", "test_value") == OK
+        await glide_client.update_connection_password(
             NEW_PASSWORD, immediate_auth=False
         )
-        assert result == OK
-        # Verify that the client is still authenticated
-        assert await glide_client.set("test_key", "test_value") == OK
-        value = await glide_client.get("test_key")
-        assert value == b"test_value"
         await config_set_new_password(glide_client, NEW_PASSWORD)
         await kill_connections_tolerant(management_client)
 
-        # Wait for the client to reconnect with the new password using retry
-        # instead of a fixed sleep, which is unreliable in cluster mode under CI load
-        async def _check_reconnected_after_first_kill():
+        async def _reconnected():
             try:
-                val = await glide_client.get("test_key")
-                return val == b"test_value"
+                return await glide_client.get("test_key") == b"test_value"
             except Exception:
                 return False
 
         await wait_for(
-            _check_reconnected_after_first_kill,
-            "Client failed to reconnect with new password after first kill",
+            _reconnected,
+            "Client did not recover with stored new password after reconnect",
             timeout=reconnect_timeout,
         )
+
+    @pytest.mark.parametrize("cluster_mode", [True, False])
+    @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
+    async def test_update_connection_password_immediate_auth_survives_reconnect(
+        self,
+        glide_client: TGlideClient,
+        management_client: TGlideClient,
+        cluster_mode: bool,
+    ):
+        """
+        Property: after a password rotation and a disconnect, calling
+        update_connection_password(new, immediate_auth=True) succeeds against
+        the reconnected client, and the client continues to accept traffic
+        with the new password.
+
+        Verifies:
+          - update_connection_password(new, immediate_auth=False)
+          - server-side rotate: CONFIG SET requirepass new
+          - server-side disconnect: CLIENT KILL (one kill only)
+          - update_connection_password(new, immediate_auth=True) returns OK
+          - subsequent SET succeeds
+
+        Applies in both cluster and standalone modes.
+        """
+        reconnect_timeout = (
+            _CLUSTER_RECONNECT_TIMEOUT_SEC
+            if cluster_mode
+            else _STANDALONE_RECONNECT_TIMEOUT_SEC
+        )
+        await glide_client.update_connection_password(
+            NEW_PASSWORD, immediate_auth=False
+        )
+        await config_set_new_password(glide_client, NEW_PASSWORD)
         await kill_connections_tolerant(management_client)
 
-        # Wait for reconnection again before attempting immediate auth
-        async def _check_reconnected_after_second_kill():
+        async def _immediate_auth_after_reconnect():
             try:
-                result = await glide_client.update_connection_password(
-                    NEW_PASSWORD, immediate_auth=True
+                return (
+                    await glide_client.update_connection_password(
+                        NEW_PASSWORD, immediate_auth=True
+                    )
+                    == OK
                 )
-                return result == OK
             except Exception:
                 return False
 
         await wait_for(
-            _check_reconnected_after_second_kill,
-            "Client failed to reconnect after second kill for immediate auth",
+            _immediate_auth_after_reconnect,
+            "Client did not accept immediate-auth after reconnect",
             timeout=reconnect_timeout,
         )
-
-        # Verify that the client is still authenticated.
-        # Use a retry loop because the per-node connections (plus, in cluster mode,
-        # the topology refresh) may still be re-stabilizing right after reconnection
-        # reports success.
-        async def _verify_authenticated():
-            try:
-                return await glide_client.set("test_key", "test_value") == OK
-            except Exception:
-                return False
-
-        await wait_for(
-            _verify_authenticated,
-            "Client not fully authenticated after reconnect",
-            timeout=reconnect_timeout,
-        )
+        assert await glide_client.set("test_key", "test_value") == OK
 
     @pytest.mark.parametrize("cluster_mode", [False])
     @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])

--- a/python/tests/async_tests/test_auth.py
+++ b/python/tests/async_tests/test_auth.py
@@ -36,8 +36,8 @@ from tests.utils.utils import (
 # Reconnect budget for the disconnect-and-recover flow in
 # test_update_connection_password. In cluster mode a CLIENT KILL is fanned out
 # over AllNodes and, under heavy full-matrix CI contention, the full-cluster
-# reconnect + re-auth can take considerably longer than on an unloaded host, so
-# cluster mode gets a wider tolerance. Standalone keeps the original budget.
+# reconnect + re-auth can take considerably longer than on an unloaded host,
+# so cluster mode gets a wider tolerance.
 _CLUSTER_RECONNECT_TIMEOUT_SEC = 90
 _STANDALONE_RECONNECT_TIMEOUT_SEC = 30
 
@@ -89,6 +89,7 @@ class TestAuthCommands:
         except RequestError:
             pass
 
+    @pytest.mark.timeout(600)
     @pytest.mark.parametrize("cluster_mode", [True, False])
     @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
     async def test_update_connection_password(
@@ -154,8 +155,9 @@ class TestAuthCommands:
         )
 
         # Verify that the client is still authenticated.
-        # Use a retry loop because the connection pool may still be stabilizing
-        # right after reconnection succeeds.
+        # Use a retry loop because the per-node connections (plus, in cluster mode,
+        # the topology refresh) may still be re-stabilizing right after reconnection
+        # reports success.
         async def _verify_authenticated():
             try:
                 return await glide_client.set("test_key", "test_value") == OK

--- a/python/tests/async_tests/test_auth.py
+++ b/python/tests/async_tests/test_auth.py
@@ -152,6 +152,7 @@ class TestAuthCommands:
             "Client failed to reconnect after second kill for immediate auth",
             timeout=reconnect_timeout,
         )
+
         # Verify that the client is still authenticated.
         # Use a retry loop because the connection pool may still be stabilizing
         # right after reconnection succeeds.

--- a/python/tests/sync_tests/test_sync_auth.py
+++ b/python/tests/sync_tests/test_sync_auth.py
@@ -11,7 +11,7 @@ from glide_shared.config import (
     ServiceType,
 )
 from glide_shared.constants import OK
-from glide_shared.exceptions import RequestError
+from glide_shared.exceptions import ClosingError, RequestError
 from glide_sync.glide_client import TGlideClient
 
 from tests.constants import (
@@ -20,7 +20,12 @@ from tests.constants import (
     IAM_TEST_REGION_US_EAST_1,
     IAM_USERNAME,
 )
-from tests.sync_tests.conftest import create_sync_client
+from tests.sync_tests.conftest import (
+    _get_worker_id,
+    _sync_client_pool,
+    _sync_client_pool_lock,
+    create_sync_client,
+)
 from tests.utils.utils import (
     NEW_PASSWORD,
     USERNAME,
@@ -68,107 +73,211 @@ def create_iam_client(
     )
 
 
+def _run_sync_cleanup_ops(management_sync_client: TGlideClient):
+    """Run auth-fixture cleanup ops; return the first failure description or None."""
+    failed_op = None
+
+    try:
+        auth_client(management_sync_client, NEW_PASSWORD)
+    except (RequestError, TimeoutError, ClosingError) as e:
+        failed_op = f"auth_client: {e}"
+
+    try:
+        config_set_new_password(management_sync_client, "")
+    except (RequestError, TimeoutError, ClosingError) as e:
+        failed_op = failed_op or f"config_set_new_password(''): {e}"
+
+    try:
+        management_sync_client.update_connection_password(None)
+    except (RequestError, TimeoutError, ClosingError) as e:
+        failed_op = failed_op or f"update_connection_password(None): {e}"
+
+    try:
+        delete_acl_username_and_password(management_sync_client, USERNAME)
+    except (RequestError, TimeoutError, ClosingError) as e:
+        failed_op = failed_op or f"delete_acl_username_and_password: {e}"
+
+    return failed_op
+
+
+def _probe_sync_cluster_wedge(management_sync_client: TGlideClient) -> bool:
+    """Return True if the management client cannot reach the cluster within 5s."""
+
+    def _probe():
+        try:
+            management_sync_client.get("__auth_cleanup_probe__")
+            return True
+        except Exception:
+            return False
+
+    try:
+        sync_wait_for(_probe, "cluster probe failed after auth cleanup", timeout=5)
+        return False
+    except (TimeoutError, RequestError, ClosingError):
+        return True
+
+
+def _evict_pooled_sync_client(request) -> None:
+    """Evict the pooled glide_sync_client so subsequent tests get a fresh one."""
+    try:
+        cluster_mode = request.getfixturevalue("cluster_mode")
+        protocol = request.getfixturevalue("protocol")
+        cache_key = (_get_worker_id(), cluster_mode, protocol)
+        with _sync_client_pool_lock:
+            _sync_client_pool.pop(cache_key, None)
+    except Exception:
+        pass
+
+
 class TestSyncAuthCommands:
     """Test cases for password authentication and management"""
 
     @pytest.fixture(autouse=True, scope="function")
     def cleanup(self, request, management_sync_client: TGlideClient):
         """
-        Ensure password is reset for default user and USERNAME user is deleted after each test, regardless of test outcome.
-        This fixture runs after each test.
+        Ensure password is reset for default user and USERNAME user is deleted
+        after each test. Per-op typed excepts so a wedge at any single step is
+        attributed rather than swallowed; a wedge probe at the end evicts the
+        pooled client and fails loudly so subsequent tests don't inherit a
+        poisoned auth cluster.
         """
         yield
-        try:
-            # reset password for default user
-            auth_client(management_sync_client, NEW_PASSWORD)
-            config_set_new_password(management_sync_client, "")
-            management_sync_client.update_connection_password(None)
 
-            # delete USERNAME user
-            delete_acl_username_and_password(management_sync_client, USERNAME)
-        except RequestError:
-            pass
+        failed_op = _run_sync_cleanup_ops(management_sync_client)
+        wedge_detected = _probe_sync_cluster_wedge(management_sync_client)
 
-    @pytest.mark.timeout(600)
+        if failed_op is not None or wedge_detected:
+            _evict_pooled_sync_client(request)
+            pytest.fail(
+                f"auth cleanup left cluster wedged: failed_op={failed_op!r}, "
+                f"wedge_probe_timed_out={wedge_detected}. Requirepass state unknown; "
+                f"subsequent tests may inherit poisoned cluster."
+            )
+
     @pytest.mark.parametrize("cluster_mode", [True, False])
     @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
-    def test_sync_update_connection_password(
+    def test_sync_update_connection_password_accepts_new_password(
         self,
         glide_sync_client: TGlideClient,
         management_sync_client: TGlideClient,
         cluster_mode: bool,
     ):
         """
-        Test replacing the connection password without immediate re-authentication.
-        Verifies that:
-        1. The client can update its internal password
-        2. The client remains connected with current auth
-        3. The client can reconnect using the new password after server password change
-        This test is only for cluster mode, as standalone mode does not have a connection available handler
+        Property: update_connection_password(new, immediate_auth=False) is a
+        pure client-side state change. The client accepts the new stored
+        password without disturbing the current session (which is still
+        authenticated with the old server-side password).
+
+        Verifies:
+          1. Return value is OK.
+          2. The still-live session accepts a subsequent SET.
+          3. The still-live session accepts a subsequent GET.
+
+        No CLIENT KILL, no server-side password change. Fast, deterministic,
+        no reconnect exposure. Applies in both cluster and standalone modes.
+        """
+        assert (
+            glide_sync_client.update_connection_password(
+                NEW_PASSWORD, immediate_auth=False
+            )
+            == OK
+        )
+        assert glide_sync_client.set("test_key", "test_value") == OK
+        assert glide_sync_client.get("test_key") == b"test_value"
+
+    @pytest.mark.parametrize("cluster_mode", [True, False])
+    @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
+    def test_sync_update_connection_password_survives_reconnect(
+        self,
+        glide_sync_client: TGlideClient,
+        management_sync_client: TGlideClient,
+        cluster_mode: bool,
+    ):
+        """
+        Property: after a password rotation, if the connection is dropped and
+        the server now requires the new password, the client recovers
+        automatically using its stored new password.
+
+        Verifies:
+          - update_connection_password(new, immediate_auth=False)
+          - server-side rotate: CONFIG SET requirepass new
+          - server-side disconnect: CLIENT KILL (one kill only)
+          - client recovers: eventual GET returns the pre-kill value
+
+        Applies in both cluster and standalone modes; cluster mode uses a
+        wider reconnect budget to absorb topology refresh.
         """
         reconnect_timeout = (
             _CLUSTER_RECONNECT_TIMEOUT_SEC
             if cluster_mode
             else _STANDALONE_RECONNECT_TIMEOUT_SEC
         )
-        result = glide_sync_client.update_connection_password(
-            NEW_PASSWORD, immediate_auth=False
-        )
-        assert result == OK
-        # Verify that the client is still authenticated
         assert glide_sync_client.set("test_key", "test_value") == OK
-        value = glide_sync_client.get("test_key")
-        assert value == b"test_value"
+        glide_sync_client.update_connection_password(NEW_PASSWORD, immediate_auth=False)
         config_set_new_password(glide_sync_client, NEW_PASSWORD)
         sync_kill_connections_tolerant(management_sync_client)
 
-        # Wait for the client to reconnect with the new password using retry
-        # instead of a fixed sleep, which is unreliable in cluster mode under CI load
-        def _check_reconnected_after_first_kill():
+        def _reconnected():
             try:
-                val = glide_sync_client.get("test_key")
-                return val == b"test_value"
+                return glide_sync_client.get("test_key") == b"test_value"
             except Exception:
                 return False
 
         sync_wait_for(
-            _check_reconnected_after_first_kill,
-            "Client failed to reconnect with new password after first kill",
+            _reconnected,
+            "Client did not recover with stored new password after reconnect",
             timeout=reconnect_timeout,
         )
+
+    @pytest.mark.parametrize("cluster_mode", [True, False])
+    @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
+    def test_sync_update_connection_password_immediate_auth_survives_reconnect(
+        self,
+        glide_sync_client: TGlideClient,
+        management_sync_client: TGlideClient,
+        cluster_mode: bool,
+    ):
+        """
+        Property: after a password rotation and a disconnect, calling
+        update_connection_password(new, immediate_auth=True) succeeds against
+        the reconnected client, and the client continues to accept traffic
+        with the new password.
+
+        Verifies:
+          - update_connection_password(new, immediate_auth=False)
+          - server-side rotate: CONFIG SET requirepass new
+          - server-side disconnect: CLIENT KILL (one kill only)
+          - update_connection_password(new, immediate_auth=True) returns OK
+          - subsequent SET succeeds
+
+        Applies in both cluster and standalone modes.
+        """
+        reconnect_timeout = (
+            _CLUSTER_RECONNECT_TIMEOUT_SEC
+            if cluster_mode
+            else _STANDALONE_RECONNECT_TIMEOUT_SEC
+        )
+        glide_sync_client.update_connection_password(NEW_PASSWORD, immediate_auth=False)
+        config_set_new_password(glide_sync_client, NEW_PASSWORD)
         sync_kill_connections_tolerant(management_sync_client)
 
-        # Wait for reconnection again before attempting immediate auth
-        def _check_reconnected_after_second_kill():
+        def _immediate_auth_after_reconnect():
             try:
-                result = glide_sync_client.update_connection_password(
-                    NEW_PASSWORD, immediate_auth=True
+                return (
+                    glide_sync_client.update_connection_password(
+                        NEW_PASSWORD, immediate_auth=True
+                    )
+                    == OK
                 )
-                return result == OK
             except Exception:
                 return False
 
         sync_wait_for(
-            _check_reconnected_after_second_kill,
-            "Client failed to reconnect after second kill for immediate auth",
+            _immediate_auth_after_reconnect,
+            "Client did not accept immediate-auth after reconnect",
             timeout=reconnect_timeout,
         )
-
-        # Verify that the client is still authenticated.
-        # Use a retry loop because the per-node connections (plus, in cluster mode,
-        # the topology refresh) may still be re-stabilizing right after reconnection
-        # reports success.
-        def _verify_authenticated():
-            try:
-                return glide_sync_client.set("test_key", "test_value") == OK
-            except Exception:
-                return False
-
-        sync_wait_for(
-            _verify_authenticated,
-            "Client not fully authenticated after reconnect",
-            timeout=reconnect_timeout,
-        )
+        assert glide_sync_client.set("test_key", "test_value") == OK
 
     @pytest.mark.parametrize("cluster_mode", [False])
     @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])

--- a/python/tests/sync_tests/test_sync_auth.py
+++ b/python/tests/sync_tests/test_sync_auth.py
@@ -152,6 +152,7 @@ class TestSyncAuthCommands:
             "Client failed to reconnect after second kill for immediate auth",
             timeout=reconnect_timeout,
         )
+
         # Verify that the client is still authenticated.
         # Use a retry loop because the connection pool may still be stabilizing
         # right after reconnection succeeds.

--- a/python/tests/sync_tests/test_sync_auth.py
+++ b/python/tests/sync_tests/test_sync_auth.py
@@ -35,10 +35,10 @@ from tests.utils.utils import (
 )
 
 # Reconnect budget for the disconnect-and-recover flow in
-# test_sync_update_connection_password. In cluster mode a CLIENT KILL is fanned
-# out over AllNodes and, under heavy full-matrix CI contention, the full-cluster
-# reconnect + re-auth can take considerably longer than on an unloaded host, so
-# cluster mode gets a wider tolerance. Standalone keeps the original budget.
+# test_update_connection_password. In cluster mode a CLIENT KILL is fanned out
+# over AllNodes and, under heavy full-matrix CI contention, the full-cluster
+# reconnect + re-auth can take considerably longer than on an unloaded host,
+# so cluster mode gets a wider tolerance.
 _CLUSTER_RECONNECT_TIMEOUT_SEC = 90
 _STANDALONE_RECONNECT_TIMEOUT_SEC = 30
 
@@ -89,6 +89,7 @@ class TestSyncAuthCommands:
         except RequestError:
             pass
 
+    @pytest.mark.timeout(600)
     @pytest.mark.parametrize("cluster_mode", [True, False])
     @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
     def test_sync_update_connection_password(
@@ -154,8 +155,9 @@ class TestSyncAuthCommands:
         )
 
         # Verify that the client is still authenticated.
-        # Use a retry loop because the connection pool may still be stabilizing
-        # right after reconnection succeeds.
+        # Use a retry loop because the per-node connections (plus, in cluster mode,
+        # the topology refresh) may still be re-stabilizing right after reconnection
+        # reports success.
         def _verify_authenticated():
             try:
                 return glide_sync_client.set("test_key", "test_value") == OK

--- a/python/tests/sync_tests/test_sync_auth.py
+++ b/python/tests/sync_tests/test_sync_auth.py
@@ -39,8 +39,8 @@ from tests.utils.utils import (
 # out over AllNodes and, under heavy full-matrix CI contention, the full-cluster
 # reconnect + re-auth can take considerably longer than on an unloaded host, so
 # cluster mode gets a wider tolerance. Standalone keeps the original budget.
-_CLUSTER_RECONNECT_TIMEOUT_SEC = 60
-_STANDALONE_RECONNECT_TIMEOUT_SEC = 25
+_CLUSTER_RECONNECT_TIMEOUT_SEC = 90
+_STANDALONE_RECONNECT_TIMEOUT_SEC = 30
 
 
 def create_iam_client(
@@ -152,8 +152,20 @@ class TestSyncAuthCommands:
             "Client failed to reconnect after second kill for immediate auth",
             timeout=reconnect_timeout,
         )
-        # Verify that the client is still authenticated
-        assert glide_sync_client.set("test_key", "test_value") == OK
+        # Verify that the client is still authenticated.
+        # Use a retry loop because the connection pool may still be stabilizing
+        # right after reconnection succeeds.
+        def _verify_authenticated():
+            try:
+                return glide_sync_client.set("test_key", "test_value") == OK
+            except Exception:
+                return False
+
+        sync_wait_for(
+            _verify_authenticated,
+            "Client not fully authenticated after reconnect",
+            timeout=reconnect_timeout,
+        )
 
     @pytest.mark.parametrize("cluster_mode", [False])
     @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])

